common/utils.c

```diff
@@ -550,3 +551,87 @@ uint32_t TDNFPathFromUri(
     goto cleanup;
 }
 
+
+uint32_t
+TDNFNormalizePath(
+    const char* pszPath,
+    char** ppszNormalPath)
+{
+    uint32_t dwError = 0;
+    char* pszNormalPath = NULL;
+    const char* p = pszPath;
+    char* q;
+
+    if(IsNullOrEmptyString(pszPath) || !ppszNormalPath)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = TDNFAllocateMemory(1, strlen(pszPath) + 1, (void **)&pszNormalPath);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    q = pszNormalPath;
+
+    while(*p)
+    {
+        /* double slashes */
+        if (*p == '/' && (p > pszPath && p[-1] == '/'))
```

Comment:
This handles multiple slashes already. Only the first slash is copied (condition is false), every subsequent one is ignored (condition true).

Reply:
The look back is needed because we need to ignore a slash if the previous one is a slash too, and similar for the other conditions. I think all look backs are sufficiently protected.

Thinking about more it, another way would be to set a flag that is `true` initially, then reset if `*p` did not point to a slash. Not sure if that makes it more readable though.

Reply:
> The look back is needed because we need to ignore a slash if the previous one is a slash too, and similar for the other conditions. I think all look backs are sufficiently protected.

With the suggested code (`while (p[0] == '/' && p[1] == '/') p++`), the idea is to keep the cursor (p) at the last char of interest (this would be a loop invariant). So in case of '///////' p would point to the last '/'. With that assumption, rest of the code can just look ahead. This works well with '..' handling too (`p[1] == '.' && p[2] == '.'`). This way all movement of the cursor will be at the top s when you are done with a section of text, a `continue` should get you back in the loop meeting all the loop invariant again.

Reply:
Thanks Siddharth! I reworked the code, and it indeed looks much cleaner this way. There is no looking back any more.